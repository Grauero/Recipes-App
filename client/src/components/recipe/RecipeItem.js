import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import posed from 'react-pose';

const AnimatedListItem = posed.li({
  shown: { opacity: 1 },
  hidden: { opacity: 0 }
});

const RecipeItem = ({
  _id, imageUrl, name, category
}) => (
  <AnimatedListItem
    className="card"
    style={{ background: `url(${imageUrl}) no-repeat center center / cover` }}
  >
    <span className={category}>{category}</span>
    <div className="card-text">
      <Link to={`/recipes/${_id}`}>
        <h4>{name}</h4>
      </Link>
    </div>
  </AnimatedListItem>
);

RecipeItem.propTypes = {
  _id: PropTypes.string.isRequired,
  imageUrl: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired
};

export default RecipeItem;
