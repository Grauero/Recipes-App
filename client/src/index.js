import React, { Fragment, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import {
  BrowserRouter as Router, Route, Switch, Redirect
} from 'react-router-dom';
import ApolloClient from 'apollo-boost';
import { ApolloProvider } from 'react-apollo';
import * as serviceWorker from './serviceWorker';

import App from './components/App';
import Navbar from './components/layouts/Navbar';
import Spinner from './components/common/Spinner';
import withSession from './components/utils/withSession';
import './index.css';

const Search = lazy(() => import('./components/recipe/Search'));
const Signin = lazy(() => import('./components/auth/Signin'));
const Signup = lazy(() => import('./components/auth/Signup'));
const AddRecipe = lazy(() => import('./components/recipe/AddRecipe'));
const RecipePage = lazy(() => import('./components/recipe/RecipePage'));
const Profile = lazy(() => import('./components/profile/Profile'));

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_SERVER,
  fetchOptions: {
    credentials: true
  },
  request: (operation) => {
    const token = localStorage.getItem('token');

    operation.setContext({
      headers: {
        authorization: token
      }
    });
  },
  onError: ({ networkError }) => {
    if (networkError) {
      console.log('Network Error', networkError);
    }
  }
});

const Root = ({ refetch, session }) => (
  <Router>
    <Fragment>
      <Navbar session={session} />
      <Switch>
        <Route exact path="/" component={App} />
        <Route path="/search" component={Search} />
        <Route path="/signin" render={() => <Signin refetch={refetch} />} />
        <Route path="/signup" render={() => <Signup refetch={refetch} />} />
        <Route path="/recipe/add" render={() => <AddRecipe session={session} />} />
        <Route path="/recipes/:_id" component={RecipePage} />
        <Route path="/profile" render={() => <Profile session={session} />} />
        <Redirect to="/" />
      </Switch>
    </Fragment>
  </Router>
);

Root.propTypes = {
  refetch: PropTypes.func.isRequired,
  session: PropTypes.instanceOf(Object).isRequired
};

const RootWithSession = withSession(Root);

ReactDOM.render(
  <ApolloProvider client={client}>
    <Suspense fallback={<div className="fallback"><Spinner /></div>}>
      <RootWithSession />
    </Suspense>
  </ApolloProvider>,
  document.getElementById('root')
);

serviceWorker.unregister();
