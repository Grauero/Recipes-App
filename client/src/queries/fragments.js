import { gql } from 'apollo-boost';

const recipeFragments = {
  recipe: gql`
    fragment CompleteRecipe on Recipe {
      _id
      name
      imageUrl
      category
      description
      instructions
      createdDate
      likes
      username
    }
  `,
  like: gql`
    fragment LikeRecipe on Recipe {
      _id
      likes
    }
  `
};

export default recipeFragments;
