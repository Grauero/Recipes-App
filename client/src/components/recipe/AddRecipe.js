import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { withRouter } from 'react-router-dom';
import CKEditor from 'react-ckeditor-component';

import Error from '../error/Error';
import withAuth, { isAuth } from '../utils/withAuth';
import { ADD_RECIPE, GET_ALL_RECIPES, GET_USER_RECIPES } from '../../queries';

const initialState = {
  name: '',
  imageUrl: '',
  instructions: '',
  category: 'Breakfast',
  description: '',
  username: ''
};

class AddRecipe extends Component {
  state = { ...initialState };

  componentDidMount() {
    this.setState({
      username: this.props.session.getCurrentUser.username
    });
  }

  clearState = () => this.setState({ ...initialState });

  handleChange = e => this.setState({ [e.target.name]: e.target.value });

  handleEditorChange = (e) => {
    const newContent = e.editor.getData();
    this.setState({ instructions: newContent });
  };

  handleSubmit = async (e, addRecipe) => {
    e.preventDefault();

    await addRecipe();

    this.clearState();
    this.props.history.push('/');
  };

  validateForm = () => {
    const {
      name, imageUrl, instructions, category, description
    } = this.state;
    const isInvalid = !name || !imageUrl || !instructions || !category || !description;

    return isInvalid;
  };

  updateCache = (cache, { data: { addRecipe } }) => {
    const { getAllRecipes } = cache.readQuery({ query: GET_ALL_RECIPES });

    cache.writeQuery({
      query: GET_ALL_RECIPES,
      data: {
        getAllRecipes: [addRecipe, ...getAllRecipes]
      }
    });
  };

  render() {
    const {
      name, imageUrl, instructions, category, description, username
    } = this.state;

    return (
      <Mutation
        mutation={ADD_RECIPE}
        variables={{
          name,
          imageUrl,
          instructions,
          category,
          description,
          username
        }}
        refetchQueries={() => [{ query: GET_USER_RECIPES, variables: { username } }]}
        update={this.updateCache}
      >
        {(addRecipe, { loading, error }) => (
          <div className="App">
            <h2 className="App">Add Recipe</h2>
            <form className="form" onSubmit={e => this.handleSubmit(e, addRecipe)}>
              <input
                type="text"
                name="name"
                placeholder="Recipe Name"
                value={name}
                onChange={this.handleChange}
              />

              <input
                type="text"
                name="imageUrl"
                placeholder="Recipe Image"
                value={imageUrl}
                onChange={this.handleChange}
              />

              <select name="category" value={category} onChange={this.handleChange}>
                <option value="Breakfast">Breakfast</option>
                <option value="Lunch">Lunch</option>
                <option value="Dinner">Dinner</option>
                <option value="Snack">Snack</option>
              </select>

              <input
                type="text"
                name="description"
                placeholder="Add description"
                value={description}
                onChange={this.handleChange}
              />

              <label htmlFor="instructions">Add Instructions</label>
              <CKEditor
                name="instructions"
                content={instructions}
                events={{ change: this.handleEditorChange }}
              />
              {/* <textarea
                name="instructions"
                placeholder="Add instructions"
                value={instructions}
                onChange={this.handleChange}
              /> */}

              <button
                disabled={loading || this.validateForm()}
                type="submit"
                className="button-primary"
              >
                Submit
              </button>
              {error && <Error error={error} />}
            </form>
          </div>
        )}
      </Mutation>
    );
  }
}

AddRecipe.propTypes = {
  session: PropTypes.shape({
    getCurrentUser: PropTypes.shape({
      username: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default withAuth(isAuth)(withRouter(AddRecipe));
