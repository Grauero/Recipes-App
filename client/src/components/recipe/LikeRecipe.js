import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';

import withSession from '../utils/withSession';
import { LIKE_RECIPE } from '../../queries';

class LikeRecipe extends Component {
  state = {
    username: '',
    liked: false
  };

  componentDidMount() {
    if (this.props.session.getCurrentUser) {
      const { username, favorites } = this.props.session.getCurrentUser;
      const { _id } = this.props;
      const prevLiked = favorites.findIndex(favorite => favorite._id === _id) > -1;

      this.setState({ username, liked: prevLiked });
    }
  }

  handleClick = (likeRecipe) => {
    this.setState(
      prevState => ({
        liked: !prevState.liked
      }),
      () => this.handleLike(likeRecipe)
    );
  };

  handleLike = async (likeRecipe) => {
    // like
    if (this.state.liked) {
      await likeRecipe();
      await this.props.refetch();
    }

    // unlike
  };

  render() {
    const { username, liked } = this.state;
    const { _id } = this.props;

    return (
      <Mutation mutation={LIKE_RECIPE} variables={{ _id, username }}>
        {likeRecipe =>
          username && (
            <button type="button" onCLick={() => this.handleClick(likeRecipe)}>
              {liked ? 'Liked' : 'Like'}
            </button>
          )
        }
      </Mutation>
    );
  }
}

LikeRecipe.propTypes = {
  session: PropTypes.shape({
    getCurrentUser: PropTypes.shape({
      username: PropTypes.string.isRequired,
      favorites: PropTypes.arrayOf(PropTypes.string).isRequired
    }).isRequired
  }).isRequired,
  _id: PropTypes.string.isRequired,
  refetch: PropTypes.func.isRequired
};

export default withSession(LikeRecipe);
