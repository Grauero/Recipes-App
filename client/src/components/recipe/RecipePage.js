import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { Query } from 'react-apollo';

import LikeRecipe from './LikeRecipe';
import Spinner from '../common/Spinner';
import Error from '../common/Error';
import { GET_RECIPE } from '../../queries';

const RecipePage = ({ match }) => {
  const { _id } = match.params;

  return (
    <Query query={GET_RECIPE} variables={{ _id }}>
      {({ data, loading, error }) => {
        if (loading) return <Spinner />;
        if (error) return <Error error={error} />;

        const {
          name, imageUrl, category, description, likes, username
        } = data.getRecipe;

        return (
          <div className="App">
            <div
              className="recipe-image"
              style={{
                background: `url(${imageUrl}) no-repeat center center / cover`
              }}
            />

            <div className="recipe">
              <div className="recipe-header">
                <h2 className="recipe-name">
                  <strong>{name}</strong>
                </h2>
                <h5>
                  <strong>{category}</strong>
                </h5>
                <p>
                  Created by <strong>{username}</strong>
                </p>
                <p>
                  {likes}{' '}
                  <span role="img" aria-label="heart">
                    ❤️
                  </span>
                </p>
              </div>

              <blockquote className="recipe-description">{description}</blockquote>

              <h3 className="recipe-instructions__title">Instructions</h3>
              <div
                className="recipe-instructions"
                dangerouslySetInnerHTML={{ __html: data.getRecipe.instructions }}
              />
              <LikeRecipe _id={_id} />
            </div>
          </div>
        );
      }}
    </Query>
  );
};

RecipePage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      _id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};

export default withRouter(RecipePage);
export { RecipePage };
