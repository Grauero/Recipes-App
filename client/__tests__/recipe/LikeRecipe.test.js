import React from 'react';
import { mount } from 'enzyme';
// import { BrowserRouter } from 'react-router-dom';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';

import { LikeRecipe } from '../../src/components/recipe/LikeRecipe';
import { LIKE_RECIPE, UNLIKE_RECIPE } from '../../src/queries';

const initialState = { username: '', liked: false };
const props = {
  session: {
    getCurrentUser: { username: 'username', favorites: [{ _id: '_id' }] }
  },
  _id: '_id',
  refetch: jest.fn()
};

const mocks = [
  {
    request: {
      query: UNLIKE_RECIPE,
      variables: {
        _id: props._id,
        username: initialState.username
      }
    },
    result: {
      data: {
        unlikeRecipe: { _id: 'id', likes: 0 }
      }
    }
  },
  {
    request: {
      query: LIKE_RECIPE,
      variables: {
        _id: props._id,
        username: initialState.username
      }
    },
    result: {
      data: {
        likeRecipe: { _id: 'id', likes: 0 }
      }
    }
  }
];
let component;
let wrapper;

beforeEach(() => {
  wrapper = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <LikeRecipe {...props} />
    </MockedProvider>
  );

  component = wrapper.find(LikeRecipe);
  component.setState(initialState);
});

afterEach(() => {
  wrapper.unmount();
});

it('renders Unlike button if user already liked the recipe', () => {
  expect(toJSON(component.find('button'))).toMatchSnapshot();
});

it('toggle liked state if user already liked the recipe', async () => {
  const prevState = component.state();
  component.find('button').simulate('click');
  await wait(0);

  expect(prevState.liked).not.toBe(component.state().liked);
  expect(props.refetch).toHaveBeenCalled();
  expect(props.refetch).toHaveBeenCalledTimes(1);
});

it('renders Like button if user did NOT liked the recipe', () => {
  const props = {
    session: {
      getCurrentUser: { username: 'username', favorites: [{ _id: '' }] }
    },
    _id: '_id',
    refetch: jest.fn()
  };
  wrapper = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <LikeRecipe {...props} />
    </MockedProvider>
  );

  component = wrapper.find(LikeRecipe);
  component.setState(initialState);

  expect(toJSON(component.find('button'))).toMatchSnapshot();
});

it('toggle liked state if user already liked the recipe', async () => {
  const props = {
    session: {
      getCurrentUser: { username: 'username', favorites: [{ _id: '' }] }
    },
    _id: '_id',
    refetch: jest.fn()
  };
  wrapper = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <LikeRecipe {...props} />
    </MockedProvider>
  );
  component = wrapper.find(LikeRecipe);
  component.setState(initialState);

  const prevState = component.state();
  component.find('button').simulate('click');
  await wait(0);

  expect(prevState.liked).not.toBe(component.state().liked);
  expect(props.refetch).toHaveBeenCalled();
  expect(props.refetch).toHaveBeenCalledTimes(1);
});
