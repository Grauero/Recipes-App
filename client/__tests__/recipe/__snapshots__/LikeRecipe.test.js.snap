// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Like button if user did NOT liked the recipe 1`] = `
<button
  className="like-button"
  onClick={[Function]}
  type="button"
>
  Like
</button>
`;

exports[`renders Unlike button if user already liked the recipe 1`] = `
<button
  className="like-button"
  onClick={[Function]}
  type="button"
>
  Unlike
</button>
`;
